ksql-common/src/test/java/io/confluent/ksql/util/timestamp/StringToTimestampParserTest.java

```diff
@@ -0,0 +1,197 @@
+package io.confluent.ksql.util.timestamp;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.equalTo;
+
+import java.sql.Timestamp;
+import java.time.LocalDateTime;
+import java.time.ZoneId;
+import java.time.ZonedDateTime;
+import org.junit.Test;
+
+public class StringToTimestampParserTest {
+
+  @Test
+  public void shouldParseFullTimestamp() {
+    // Given:
+    final StringToTimestampParser parser = new StringToTimestampParser("yyyy/MM/dd:HH");
+
+    // When:
+    final long ts = parser.parse("1605/11/05:12");
+
+    // Then:
+    assertThat(ts, equalTo(toMillis(LocalDateTime.of(1605, 11, 5, 12, 0))));
+  }
+
+  @Test
+  public void shouldParseFullTimestampWithMinutes() {
+    // Given:
+    final StringToTimestampParser parser = new StringToTimestampParser("yyyy/MM/dd HH:mm");
+
+    // When:
+    final long ts = parser.parse("1605/11/05 12:10");
+
+    // Then:
+    assertThat(ts, equalTo(toMillis(LocalDateTime.of(1605, 11, 5, 12, 10))));
+  }
+
+  @Test
+  public void shouldParseWithOptionalElements() {
+    // Given:
+    final StringToTimestampParser parser = new StringToTimestampParser("yyyy/MM/dd HH[:mm]");
+
+    // When:
+    final long ts = parser.parse("1605/11/05 12");
+
+    // Then:
+    assertThat(ts, equalTo(toMillis(LocalDateTime.of(1605, 11, 5, 12, 0))));
+  }
+
+  @Test
+  public void shouldParseWithOptionalDefaultedFields() {
+    // Given:
+    final StringToTimestampParser parser = new StringToTimestampParser("yyyy/MM[/dd]");
+
+    // When:
+    final long ts = parser.parse("1605/11/05");
+
+    // Then:
+    assertThat(ts, equalTo(toMillis(LocalDateTime.of(1605, 11, 5, 0, 0))));
+  }
+
+  @Test
+  public void shouldParseClockHourOfDay() {
+    // Given:
+    final StringToTimestampParser parser = new StringToTimestampParser("yyyy/MM/dd[ k]");
+
+    // When:
+    final long ts = parser.parse("1605/11/01 2");
+
+    // Then:
+    assertThat(ts, equalTo(toMillis(LocalDateTime.of(1605, 11, 1, 2, 0))));
+  }
+
+  @Test
+  public void shouldParseWithTimeZone() {
+    // Given:
+    final StringToTimestampParser parser = new StringToTimestampParser("yyyy/MM/dd HH z");
+
+    // When:
+    final long ts = parser.parse("1605/11/01 12 UTC");
+
+    // Then:
+    assertThat(ts, equalTo(
+        toMillis(ZonedDateTime.of(1605, 11, 1, 12, 0, 0, 0, ZoneId.of("UTC"))
+            .withZoneSameInstant(ZoneId.systemDefault())
+            .toLocalDateTime())));
+  }
+
+  @Test
+  public void shouldResolveDefaultsWithNoFormat() {
+    // Given:
+    final StringToTimestampParser parser = new StringToTimestampParser("");
+
+    // When:
+    final long ts = parser.parse("");
+
+    // Then:
+    assertThat(ts, equalTo(toMillis(LocalDateTime.of(1970, 1, 1, 0, 0))));
+  }
+
+  @Test
+  public void shouldResolveDefaultsWithPartialFormat() {
+    // Given:
+    final StringToTimestampParser parser = new StringToTimestampParser("yyyy");
```

Comment:
All of these tests include the year. Is it valid to not supply a year, and if one isn't provided does it default to 1970?  If it does, then it seems like a strange default.  I guess the benefit is that its constant, but maybe a more useful default is the current year?  e.g. if I parse a date of `2nd Feb` today should it parse as `2019-02-02` or `1970-02-02`?

Reply:
All of the tests, except for the one you commented on above! 😉 It does default to 1970, and that is the Unix epoch (Thursday Jan 1st, 1970) so I think it's not an insane default (sane is indeed a different question). I'm not opposed to creating it with the current year, but I can imagine some tricky scenarios where during some weird edge cases we create `year+1/dec/31` and that can get kinda bad. 

Since this is keeping the legacy behavior, and no one has complained, maybe we leave it as is now and in a future PR we can change the default since I'm not sure it'll be as trivial as "just do it" TM.

Reply:
Actual now that I think of it, I prefer the default behavior. Automated tests that run would like to not have changing outputs based on the system time and if the date is important to the end user, I think we should encourage them to specify a full date structure for predictability.

Reply:
Yeah, I hear you.  Consistent results, regardless of processing time, is a good thing for sure.

I just think having a partial date that defaults to 1970 probably means such partial dates are pretty useless!

Is there a format string users can use to provide an alternative year within the format string? e.g. is it valid to pass `"2019/MM/dd HH:mm"` to change the default year?

Reply:
> I just think having a partial date that defaults to 1970 probably means such partial dates are pretty useless!

I totally agree. And I see a use case for sending something like `Feb 2` and assuming current year (though I don't really see any good use case for other partial dates).

> Is there a format string users can use to provide an alternative year within the format string? e.g. is it valid to pass "2019/MM/dd HH:mm" to change the default year?

I don't think something like this is possible since the parser doesn't know what 2019 is out of context.

----

I think my current line of thought is that this PR makes things better, but that we really need to introduce a Date object if we want to get this functionality "right".